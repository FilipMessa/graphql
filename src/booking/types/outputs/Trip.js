// @flow

import { GraphQLObjectType, GraphQLList, GraphQLInt } from 'graphql';
import idx from 'idx';

import Leg from '../../../flight/types/outputs/Leg';
import GraphQLRouteStop from '../../../flight/types/outputs/RouteStop';
import type { DepartureArrival, Leg as LegType } from '../../../flight/Flight';
import FlightDurationInMinutes, {
  getDurationInMinutes,
} from '../../../flight/resolvers/FlightDuration';

export type TripData = {
  departure: DepartureArrival,
  arrival: DepartureArrival,
  legs: LegType[],
  bid: number,
};

export default new GraphQLObjectType({
  name: 'Trip',
  description:
    'Single travel from origin to destination, with possible stopovers.',
  fields: {
    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: TripData): DepartureArrival => departure,
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: TripData): DepartureArrival => arrival,
    },

    duration: {
      type: GraphQLInt,
      description: 'Trip duration in minutes.',
      resolve: ({ departure, arrival }: TripData): ?number => {
        return FlightDurationInMinutes(departure, arrival);
      },
    },

    legs: {
      type: new GraphQLList(Leg),
      resolve: ({ legs }: TripData): LegType[] =>
        legs.map((leg, index) => ({
          ...leg,
          stopoverDuration: getStopoverDuration(leg, legs[index - 1]),
        })),
    },
  },
});

const getStopoverDuration = (
  leg: LegType,
  previousLeg: ?LegType,
): number | null => {
  if (previousLeg == null) {
    return null;
  }

  const currentLegDeparture = idx(leg.departure, _ => _.when.utc);
  const previousLegArrival = idx(previousLeg.arrival, _ => _.when.utc);

  if (currentLegDeparture == null || previousLegArrival == null) {
    return null;
  }

  return getDurationInMinutes(
    new Date(previousLegArrival),
    new Date(currentLegDeparture),
  );
};
