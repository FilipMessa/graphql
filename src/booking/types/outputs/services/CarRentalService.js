// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
