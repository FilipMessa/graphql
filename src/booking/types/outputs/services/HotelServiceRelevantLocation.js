// @flow

import { GraphQLObjectType } from 'graphql';
import idx from 'idx';

import Location from '../../../../location/types/outputs/Location';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';
import HotelCityGraphqQLType from '../../../../hotel/types/outputs/HotelCity';

export default new GraphQLObjectType({
  name: 'HotelServiceRelevantLocation',
  fields: {
    location: {
      type: Location,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        return location;
      },
    },
    hotelCity: {
      type: HotelCityGraphqQLType,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader, locale }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        let hotelCities = [];
        const cityName = idx(location, _ => _.city.name);

        if (cityName) {
          hotelCities = await dataLoader.hotel.cities.loadByPrefix(
            cityName,
            locale.language,
          );
        }

        const lat = idx(location, _ => _.location.lat);
        const lng = idx(location, _ => _.location.lng);
        if (hotelCities.length === 0 && lat != null && lng != null) {
          /**
           * If we don't find any city quering by name, let's fall back to the coordinates of the airport
           * Note that this is most likely not going to be the destination city
           */
          hotelCities = await dataLoader.hotel.cities.loadByLatLng(lat, lng);
        }

        return hotelCities[0];
      },
    },
  },
});
