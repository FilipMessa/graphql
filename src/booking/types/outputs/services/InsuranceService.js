// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import GraphQLPassenger from '../Passenger';

export default new GraphQLObjectType({
  name: 'InsuranceService',
  fields: {
    passengers: {
      type: GraphQLList(GraphQLPassenger),
      description: 'Returns passengers allowed to buy a travel insurance.',
    },
  },
});
