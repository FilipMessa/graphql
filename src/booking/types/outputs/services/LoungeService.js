// @flow

import { GraphQLObjectType, GraphQLString, GraphQLList } from 'graphql';
import Location from '../../../../location/types/outputs/Location';
import LoungeWhiteLabelURLResolver from '../../../resolvers/LoungeWhitelabelURL';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';

type LoungeServiceInput = {|
  +iataCodes: Set<string>,
  +departureTime: Date,
|};

type LoungeServiceOutput = {|
  +iataCode: string,
  +departureTime: Date,
|};

export default new GraphQLObjectType({
  name: 'LoungeService',
  fields: {
    relevantAirports: {
      type: GraphQLList(
        new GraphQLObjectType({
          name: 'LoungeServiceRelevantAirports',
          fields: {
            whitelabelURL: {
              type: GraphQLString,
              resolve: async ({
                iataCode,
                departureTime,
              }: LoungeServiceOutput) => {
                return LoungeWhiteLabelURLResolver(iataCode, departureTime);
              },
            },

            location: {
              type: Location,
              resolve: (
                { iataCode }: LoungeServiceOutput,
                args: void,
                context: GraphqlContextType,
              ) => {
                return context.dataLoader.location.load(iataCode);
              },
            },
          },
        }),
      ),
      resolve: ({
        iataCodes,
        departureTime,
      }: LoungeServiceInput): $ReadOnlyArray<LoungeServiceOutput> => {
        return Array.from(iataCodes).map(iataCode => ({
          iataCode,
          departureTime,
        }));
      },
    },
  },
});
