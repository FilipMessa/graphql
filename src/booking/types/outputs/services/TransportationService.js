// @flow

import { GraphQLList, GraphQLObjectType } from 'graphql';

import TransportationServiceRelevantLocations from './TransportationServiceRelevantLocations';

export default new GraphQLObjectType({
  name: 'TransportationService',
  fields: {
    relevantLocations: {
      type: GraphQLList(TransportationServiceRelevantLocations),
      resolve: ancestor => {
        return ancestor;
      },
    },
  },
});
