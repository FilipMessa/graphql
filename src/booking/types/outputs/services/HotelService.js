// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
