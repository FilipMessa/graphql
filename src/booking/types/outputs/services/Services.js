// @flow

import { GraphQLObjectType } from 'graphql';
import { DateTime } from 'luxon';

import InsuranceService from './InsuranceService';
import type { BookingsItem } from '../../../Booking';

type AncestorType = {|
  +booking: BookingsItem,
|};

export default new GraphQLObjectType({
  name: 'Services',
  fields: {
    insurance: {
      type: InsuranceService,
      description: 'Travel insurance - returns "null" if not available.',
      resolve: async ({ booking }: AncestorType) => {
        // booking must be CONFIRMED
        if (booking.status !== 'confirmed') {
          return null;
        }

        if (!booking.departure.when) {
          return null;
        }

        // departure must be at least 48 hours from now
        const diff = DateTime.fromJSDate(booking.departure.when.utc)
          .diff(DateTime.fromJSDate(new Date()), 'hours')
          .toObject().hours;

        if (diff < 48) {
          return null;
        }

        // we do not provide insurance for US nationality (need to check all PAX)
        const validPax = booking.passengers.filter(
          pap => pap.nationality !== 'us',
        );
        if (validPax.length === 0) {
          return null;
        }

        return {
          passengers: validPax,
        };
      },
    },
  },
});
