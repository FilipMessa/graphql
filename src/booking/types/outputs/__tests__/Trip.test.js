// @flow

import Trip from '../Trip';
import { evaluateResolver } from '../../../../common/services/TestingTools';

const fields = Trip.getFields();

it('Trip type should have valid fields', () => {
  expect(fields).toMatchSnapshot();
});

describe('Legs resolver', () => {
  it('should set stopoverDuration null for first leg', () => {
    const result = evaluateResolver(fields.legs, {
      legs: [{ test: '' }],
    });

    expect(result[0].stopoverDuration).toBe(null);
  });

  it('should calculate duration correctly for other legs', () => {
    const result = evaluateResolver(fields.legs, {
      legs: [
        { arrival: { when: { utc: 500 } } },
        { departure: { when: { utc: 500 + 120 * 1000 * 60 } } },
      ],
    });

    expect(result[1].stopoverDuration).toBe(120);
  });
});
