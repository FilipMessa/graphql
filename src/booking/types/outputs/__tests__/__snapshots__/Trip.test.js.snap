// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trip type should have valid fields 1`] = `
Object {
  "arrival": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "arrival",
    "resolve": [Function],
    "type": "RouteStop",
  },
  "departure": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "departure",
    "resolve": [Function],
    "type": "RouteStop",
  },
  "duration": Object {
    "args": Array [],
    "description": "Trip duration in minutes.",
    "isDeprecated": false,
    "name": "duration",
    "resolve": [Function],
    "type": "Int",
  },
  "legs": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "legs",
    "resolve": [Function],
    "type": "[Leg]",
  },
}
`;
